adapters/criteo/models.go

```diff
@@ -0,0 +1,333 @@
+		criteoUser.DeviceIDType = deviceType
+
+		if device.OS != "" {
+			criteoUser.DeviceOS = device.OS
+		}
+		if device.IFA != "" {
+			criteoUser.DeviceID = device.IFA
+		}
+		if device.IP != "" {
+			criteoUser.IP = device.IP
+		}
+		if device.UA != "" {
+			criteoUser.UA = device.UA
+		}
+	}
+
+	if regsExt != nil {
+		if regsExt.USPrivacy != "" {
+			criteoUser.UspIab = regsExt.USPrivacy // CCPA
+		}
+	}
+
+	return criteoUser
+}
+
+type criteoGdprConsent struct {
+	GdprApplies *bool  `json:"gdprapplies,omitempty"`
+	ConsentData string `json:"consentdata,omitempty"`
+}
+
+func newCriteoGdprConsent(user *openrtb2.User, regsExt *openrtb_ext.ExtRegs) (criteoGdprConsent, error) {
+	consent := criteoGdprConsent{}
+
+	if user == nil && regsExt == nil {
+		return consent, nil
+	}
+
+	if user != nil {
+		if user.Ext != nil {
+			var userExt *openrtb_ext.ExtUser
+			if err := json.Unmarshal(user.Ext, &userExt); err != nil {
+				return consent, err
+			}
+			if userExt.Consent != "" {
+				consent.ConsentData = userExt.Consent
+			}
+		}
+	}
+
+	if regsExt != nil {
+		if regsExt.GDPR != nil {
+			gdprApplies := bool((*regsExt.GDPR & 1) == 1)
+			consent.GdprApplies = &gdprApplies
+		}
+	}
+
+	return consent, nil
+}
+
+type criteoRequestSlot struct {
+	SlotID      string              `json:"slotid,omitempty"`
+	ImpID       string              `json:"impid,omitempty"`
+	ZoneID      *int64              `json:"zoneid,omitempty"`
+	NetworkID   *int64              `json:"networkid,omitempty"`
+	PlacementID string              `json:"placement,omitempty"`
+	Sizes       []criteoRequestSize `json:"sizes,omitempty"`
+}
+
+func newCriteoRequestSlots(slotIDGenerator slotIDGenerator, impressions []openrtb2.Imp) ([]criteoRequestSlot, []error) {
+	var errs []error
+
+	if len(impressions) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Bid request impressions is nil or empty",
+		}}
+	}
+
+	// TODO: Criteo slot should comes either with:
+	//   - `zoneid`
+	// OR
+	//   - `networkid`, `placementid`, `sizes`
+	//
+	// if not, criteo will reject the slot.
+	// Would be nice preventing PBS to send such slots when conditions aren't met
+
+	var criteoSlots = make([]criteoRequestSlot, int64(len(impressions)))
+
+	for i := 0; i < len(impressions); i++ {
+		criteoSlots[i] = criteoRequestSlot{}
+
+		if impressions[i].ID != "" {
+			criteoSlots[i].ImpID = impressions[i].ID
+		}
+
+		// Generating a random slot ID
+		generatedSlotID, err := slotIDGenerator.NewSlotID()
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		criteoSlots[i].SlotID = generatedSlotID
+
+		if impressions[i].Banner != nil {
+			if impressions[i].Banner.Format != nil {
+				criteoSlots[i].Sizes = make([]criteoRequestSize, len(impressions[i].Banner.Format))
+				for idx, format := range impressions[i].Banner.Format {
+					// TODO: handle properly int64 conversion
+					criteoSlots[i].Sizes[idx] = newCriteoRequestSize(int64(format.W), int64(format.H))
```

Comment:
With the update to the openrtb library, I believe you no longe need the `int64` casts.

Reply:
Indeed :)